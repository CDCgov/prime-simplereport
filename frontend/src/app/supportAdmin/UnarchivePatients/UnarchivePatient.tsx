import { useForm } from "react-hook-form";
import React, { useEffect, useState } from "react";
import moment from "moment/moment";
import { useNavigate } from "react-router-dom";

import { useDocumentTitle } from "../../utils/hooks";
import { PATIENT_TERM, PATIENT_TERM_CAP } from "../../../config/constants";
import Select from "../../commonComponents/Select";
import {
  ArchivedStatus,
  useGetOrganizationWithFacilitiesLazyQuery,
  useGetPatientsByFacilityWithOrgLazyQuery,
  useGetPatientsCountByFacilityWithOrgLazyQuery,
} from "../../../generated/graphql";
import { displayFullName } from "../../utils";
import Button from "../../commonComponents/Button/Button";
import Pagination from "../../commonComponents/Pagination";
import { useSelectedFacility } from "../../facilitySelect/useSelectedFacility";

import {
  UnarchivePatientFacility,
  UnarchivePatientOrganization,
  UnarchivePatientPatient,
} from "./UnarchivePatientContainer";

interface UnarchivePatientProps {
  organizations: UnarchivePatientOrganization[] | [];
  currentPage: number;
  selectedFacilityId: string;
  setSelectedFacilityId: (facilityId: string) => void;
  selectedOrgId: string;
  setSelectedOrgId: (orgInternalId: string) => void;
  archivedPatientsCount: undefined | number;
  setArchivedPatientsCount: (count: number | undefined) => void;
  facilities: UnarchivePatientFacility[] | [];
  setFacilities: (facilities: UnarchivePatientFacility[]) => void;
  archivedPatients: UnarchivePatientPatient[] | undefined;
  setArchivedPatients: (
    patients: UnarchivePatientPatient[] | undefined
  ) => void;
}

const displayPatientTable = (
  patients: UnarchivePatientPatient[] | undefined,
  patientsCount: number | undefined
) => {
  return patients !== undefined && patientsCount !== undefined;
};

const displayPagination = (
  patients: UnarchivePatientPatient[] | undefined,
  patientsCount: number | undefined
) => {
  if (patients && patientsCount) {
    return patientsCount > 0 && patients.length > 0;
  } else {
    return false;
  }
};

const displayInstructions = (
  selectedOrgId: string,
  patients: UnarchivePatientPatient[] | undefined,
  patientsCount: number | undefined,
  facilities: UnarchivePatientFacility[]
) => {
  if (
    !displayPatientTable(patients, patientsCount) &&
    !displayNoFacilitiesMsg(selectedOrgId, facilities)
  ) {
    return (
      <div className="width-full margin-top-2">
        Filter by organization and testing facility to display archived
        patients.
      </div>
    );
  }
};

const displayNoFacilitiesMsg = (
  selectedOrgId: string,
  facilities: UnarchivePatientFacility[]
) => {
  if (selectedOrgId && facilities?.length === 0) {
    return (
      <div className="width-full margin-top-2">
        This organization has no facilities. Select a different organization.
      </div>
    );
  }
};

const getOrgExternalIdWithInternalId = (
  orgs: UnarchivePatientOrganization[],
  internalId: string
) => {
  let org = orgs.find((org) => org.internalId === internalId);
  return org?.externalId;
};
const UnarchivePatient = ({
  organizations,
  currentPage,
  selectedFacilityId,
  setSelectedFacilityId,
  selectedOrgId,
  setSelectedOrgId,
  archivedPatientsCount,
  setArchivedPatientsCount,
  facilities,
  setFacilities,
  archivedPatients,
  setArchivedPatients,
}: UnarchivePatientProps) => {
  const pageTitle = `Unarchive ${PATIENT_TERM}`;
  useDocumentTitle(pageTitle);
  const unarchivePatientPageUrl = "/admin/unarchive-patient";
  const entriesPerPage = 20;

  const [isLoadingCount, setIsLoadingCount] = useState(false);
  const [isLoadingPatients, setIsLoadingPatients] = useState(false);

  const [activeFacility] = useSelectedFacility();
  const navigate = useNavigate();

  const [queryGetOrgWithFacilities] = useGetOrganizationWithFacilitiesLazyQuery(
    {
      fetchPolicy: "no-cache",
    }
  );
  const [queryGetPatientsByFacilityWithOrg] =
    useGetPatientsByFacilityWithOrgLazyQuery({
      fetchPolicy: "no-cache",
    });

  const [queryGetPatientsCountByFacilityWithOrg] =
    useGetPatientsCountByFacilityWithOrgLazyQuery({
      fetchPolicy: "no-cache",
    });

  const {
    getValues,
    handleSubmit,
    register,
    watch,
    formState: { errors },
  } = useForm({});

  const formCurrentValues = watch();

  useEffect(() => {
    fetchAndSetArchivedPatients(
      selectedOrgId,
      selectedFacilityId,
      organizations
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentPage]);

  const onOrgChange = async (selectedOrgInternalId: string) => {
    setFacilities([]);
    setArchivedPatients(undefined);
    setArchivedPatientsCount(undefined);
    if (selectedOrgInternalId) {
      setSelectedOrgId(selectedOrgInternalId);
      let res = await queryGetOrgWithFacilities({
        variables: { id: selectedOrgInternalId },
      });
      let facilities = res.data?.organization?.facilities
        ? res.data?.organization?.facilities
        : [];
      if (facilities !== undefined && facilities.length > 0) {
        facilities.sort((a, b) => (a.name > b.name ? 1 : -1));
      }
      setFacilities(facilities);
    }
  };

  const onFacilityChange = (facilityId: string) => {
    if (facilityId === "") {
      setArchivedPatientsCount(undefined);
      setArchivedPatients(undefined);
    }
    setSelectedFacilityId(facilityId);
  };

  const fetchAndSetArchivedPatientsCount = async (
    orgInternalId: string,
    facilityId: string,
    organizations: UnarchivePatientOrganization[]
  ) => {
    if (orgInternalId && facilityId && organizations) {
      setIsLoadingCount(true);
      let orgExternalId = getOrgExternalIdWithInternalId(
        organizations,
        orgInternalId
      );
      if (orgExternalId) {
        let { data } = await queryGetPatientsCountByFacilityWithOrg({
          variables: {
            facilityId: facilityId,
            archivedStatus: ArchivedStatus.Archived,
            orgExternalId: orgExternalId,
          },
        });
        let newArchivedPatientsCount = data?.patientsCount || 0;
        setArchivedPatientsCount(newArchivedPatientsCount);
      }
      setIsLoadingCount(false);
    }
  };

  const fetchAndSetArchivedPatients = async (
    orgInternalId: string,
    facilityId: string,
    organizations: UnarchivePatientOrganization[]
  ) => {
    if (orgInternalId && facilityId && organizations) {
      setIsLoadingPatients(true);
      let orgExternalId = getOrgExternalIdWithInternalId(
        organizations,
        orgInternalId
      );
      if (orgExternalId) {
        let { data } = await queryGetPatientsByFacilityWithOrg({
          variables: {
            facilityId: facilityId,
            pageNumber: currentPage - 1,
            pageSize: entriesPerPage,
            archivedStatus: ArchivedStatus.Archived,
            orgExternalId: orgExternalId,
          },
        });
        let newArchivedPatients: any[] = data?.patients || [];
        setArchivedPatients(newArchivedPatients);
        setIsLoadingPatients(false);
      }
    }
  };

  const onSubmit = async () => {
    const values = getValues();
    let orgId = values.organization;
    let facilityId = values.facility;
    if (orgId && facilityId) {
      let searchParams = "";
      if (activeFacility?.id) {
        searchParams = `?facility=${activeFacility.id}`;
      }
      await fetchAndSetArchivedPatients(orgId, facilityId, organizations);
      await fetchAndSetArchivedPatientsCount(orgId, facilityId, organizations);
      // clear pagination on search submit
      navigate(`${unarchivePatientPageUrl}${searchParams}`);
    }
  };

  const patientRows = (
    archivedPatients: UnarchivePatientPatient[] | undefined,
    archivedPatientsCount: number | undefined
  ) => {
    if (displayPagination(archivedPatients, archivedPatientsCount)) {
      return archivedPatients?.map((patient: UnarchivePatientPatient) => {
        let fullName = displayFullName(
          patient.firstName,
          patient.middleName,
          patient.lastName
        );

        return (
          <tr key={patient.internalId} data-testid="sr-archived-patient-row">
            <td>{fullName}</td>
            <td>{moment(patient.birthDate).format("MM/DD/yyyy")}</td>
            <td>
              {patient.facility ? patient.facility.name : "All facilities"}
            </td>
            <td>
              {/*Need to update in #6064*/}
              <Button
                type="button"
                label="Unarchive"
                aria-label={`Unarchive ${fullName}`}
              />
            </td>
          </tr>
        );
      });
    } else {
      return (
        <tr>
          <td colSpan={5}>No results</td>
        </tr>
      );
    }
  };

  return (
    <div className="prime-home flex-1">
      <div className="grid-container">
        <div className="grid-row">
          <div className="prime-container card-container">
            <div className="usa-card__header">
              <div className="display-flex flex-align-center">
                <h1 className="font-sans-lg margin-y-0">{pageTitle}</h1>
                <span className="sr-showing-patients-on-page margin-left-4">
                  {isLoadingCount && "Loading..."}
                  {displayPagination(
                    archivedPatients,
                    archivedPatientsCount
                  ) && (
                    <>
                      Showing {(currentPage - 1) * entriesPerPage + 1}-
                      {Math.min(
                        entriesPerPage * currentPage,
                        archivedPatientsCount ?? 0
                      )}{" "}
                      of {archivedPatientsCount}
                    </>
                  )}
                </span>
              </div>
            </div>
            <div className="bg-base-lightest">
              <form
                className="display-flex grid-row grid-gap flex-row flex-align-end padding-x-3 padding-y-2"
                role="search"
                onSubmit={handleSubmit(onSubmit)}
              >
                <div className={"desktop:grid-col-4 mobile:grid-col-12"}>
                  <Select
                    label="Organization"
                    name="organization"
                    value={formCurrentValues.organization}
                    defaultOption={"Select an organization"}
                    defaultSelect={true}
                    required={true}
                    options={organizations.map((o) => ({
                      label: o.name,
                      value: o.internalId,
                    }))}
                    registrationProps={register("organization", {
                      onChange: (e) => onOrgChange(e.target.value),
                      required: true,
                    })}
                    validationStatus={
                      errors?.organization?.type === "required"
                        ? "error"
                        : undefined
                    }
                    errorMessage="Organization is required"
                  />
                </div>
                <div className={"desktop:grid-col-4 mobile:grid-col-12"}>
                  <Select
                    label="Testing facility"
                    name="facilities"
                    defaultOption={"Select a testing facility"}
                    defaultSelect={true}
                    required={true}
                    value={formCurrentValues.facility}
                    options={facilities.map((o) => ({
                      label: o.name,
                      value: o.id,
                    }))}
                    registrationProps={register("facility", {
                      onChange: (e) => onFacilityChange(e.target.value),
                      required: true,
                    })}
                    validationStatus={
                      errors?.facility?.type === "required"
                        ? "error"
                        : undefined
                    }
                    errorMessage="Testing facility is required"
                  />
                </div>
                <div className={"mobile:margin-top-3"}>
                  <Button type="submit" label="Search" />
                </div>
              </form>
            </div>
            <div className="usa-card__body sr-patient-list">
              {isLoadingPatients && "Loading..."}
              {displayPatientTable(archivedPatients, archivedPatientsCount) && (
                <table className="usa-table usa-table--borderless width-full">
                  <thead>
                    <tr>
                      <th scope="col">{PATIENT_TERM_CAP}</th>
                      <th scope="col">Date of birth</th>
                      <th scope="col">Testing facility</th>
                      <th scope="col">Action</th>
                    </tr>
                  </thead>
                  <tbody aria-live="polite">
                    {patientRows(archivedPatients, archivedPatientsCount)}
                  </tbody>
                </table>
              )}
              {displayInstructions(
                formCurrentValues.organization,
                archivedPatients,
                archivedPatientsCount,
                facilities
              )}
              {displayNoFacilitiesMsg(
                formCurrentValues.organization,
                facilities
              )}
            </div>
            {displayPagination(archivedPatients, archivedPatientsCount) && (
              <div className="usa-card__footer">
                <Pagination
                  baseRoute={unarchivePatientPageUrl}
                  currentPage={currentPage}
                  entriesPerPage={entriesPerPage}
                  totalEntries={archivedPatientsCount ?? 0}
                />
              </div>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};

export default UnarchivePatient;
