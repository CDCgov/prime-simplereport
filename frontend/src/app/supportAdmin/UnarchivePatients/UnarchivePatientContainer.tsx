import React, { useState } from "react";
import { useParams } from "react-router-dom";

import { useGetOrganizationsQuery } from "../../../generated/graphql";
import { LoadingCard } from "../../commonComponents/LoadingCard/LoadingCard";
import { showError } from "../../utils/srToast";

import UnarchivePatients from "./UnarchivePatient";

export type UnarchivePatientOrganization = {
  internalId: string;
  externalId: string;
  name: string;
  facilities?: {
    id: string;
    name: string;
  }[];
};

export type UnarchivePatientFacility = {
  id: string;
  name: string;
};

export type UnarchivePatientPatient = {
  internalId: string;
  firstName: string;
  lastName: string;
  middleName: string;
  birthDate: string;
  isDeleted: boolean;
  facility: UnarchivePatientFacility | null;
};

const getSortedOrgs = (orgs: UnarchivePatientOrganization[] | undefined) => {
  return orgs ? orgs.sort((a, b) => (a.name > b.name ? 1 : -1)) : [];
};

const UnarchivePatientContainer = () => {
  const { pageNumber } = useParams();
  const currentPage = pageNumber ? +pageNumber : 1;

  const { data, loading, error } = useGetOrganizationsQuery({
    fetchPolicy: "no-cache",
    variables: { identityVerified: true },
  });
  const sortedOrgs = getSortedOrgs(data?.organizations);
  const [selectedOrgId, setSelectedOrgId] = useState<string>("");
  const [selectedFacilityId, setSelectedFacilityId] = useState<string>("");
  const [archivedPatientsCount, setArchivedPatientsCount] = useState<
    number | undefined
  >(undefined);
  const [archivedPatients, setArchivedPatients] = useState<
    UnarchivePatientPatient[] | undefined
  >(undefined);
  const [facilities, setFacilities] = useState<UnarchivePatientFacility[]>([]);

  if (loading) {
    return <LoadingCard message={"Loading Organizations"} />;
  }
  if (error) {
    showError(error.message, "Something went wrong");
  }

  return (
    <UnarchivePatients
      organizations={sortedOrgs}
      currentPage={currentPage}
      selectedFacilityId={selectedFacilityId}
      setSelectedFacilityId={setSelectedFacilityId}
      selectedOrgId={selectedOrgId}
      setSelectedOrgId={setSelectedOrgId}
      archivedPatients={archivedPatients}
      setArchivedPatients={setArchivedPatients}
      archivedPatientsCount={archivedPatientsCount}
      setArchivedPatientsCount={setArchivedPatientsCount}
      facilities={facilities}
      setFacilities={setFacilities}
    />
  );
};

export default UnarchivePatientContainer;
