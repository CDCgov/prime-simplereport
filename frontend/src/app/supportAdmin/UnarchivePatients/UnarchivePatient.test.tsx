import React from "react";
import {
  act,
  render,
  screen,
  waitForElementToBeRemoved,
  within,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import { MockedProvider, MockedProviderProps } from "@apollo/client/testing";
import createMockStore from "redux-mock-store";
import { MemoryRouter } from "react-router-dom";

import {
  ArchivedStatus,
  GetOrganizationWithFacilitiesDocument,
  GetPatientsByFacilityWithOrgDocument,
  GetPatientsCountByFacilityWithOrgDocument,
} from "../../../generated/graphql";

import {
  UnarchivePatientOrganization,
  UnarchivePatientFacility,
  UnarchivePatientPatient,
} from "./UnarchivePatientContainer";
import UnarchivePatient from "./UnarchivePatient";
import { selectDropdown } from "./UnarchivePatientContainer.test";

const mockFacility1: UnarchivePatientFacility = {
  id: "bc0536e-4564-4291-bbf3-0e7b0731f9e8",
  name: "Mars Facility",
};
const mockFacility2: UnarchivePatientFacility = {
  id: "d70bb3b3-96bd-40d1-a3ce-b266a7edb91d",
  name: "Jupiter Facility",
};
const mockPatient1: UnarchivePatientPatient = {
  birthDate: "1927-05-19",
  firstName: "Rod",
  internalId: "60b79a6b-5547-4d54-9cdd-e99ffef59bfc",
  isDeleted: true,
  lastName: "Gutmann",
  middleName: "",
  facility: null,
};
const mockPatient2: UnarchivePatientPatient = {
  birthDate: "1973-11-20",
  firstName: "Mia",
  internalId: "b0612367-1c82-46ad-88db-2985ac6b81ab",
  isDeleted: true,
  lastName: "Mode",
  middleName: "",
  facility: mockFacility1,
};
const mockOrg1: UnarchivePatientOrganization = {
  internalId: "f34183c4-b4c5-449f-98b0-2e02abb7aae0",
  externalId: "DC-Space-Camp-f34183c4-b4c5-449f-98b0-2e02abb7aae0",
  name: "Space Org",
  facilities: [mockFacility1, mockFacility2],
};
const mockOrgs = [mockOrg1];

const mockNavigate = jest.fn();
const mockLocation = jest.fn();

jest.mock("react-router-dom", () => {
  const original = jest.requireActual("react-router-dom");
  return {
    ...original,
    useNavigate: () => mockNavigate,
    useLocation: () => mockLocation,
  };
});
const mocks: MockedProviderProps["mocks"] = [
  {
    request: {
      query: GetOrganizationWithFacilitiesDocument,
      variables: {
        id: mockOrg1.internalId,
      },
    },
    result: {
      data: {
        organization: mockOrg1,
      },
    },
  },
  {
    request: {
      query: GetOrganizationWithFacilitiesDocument,
      variables: {
        id: mockOrg1.internalId,
      },
    },
    result: {
      data: {
        organization: mockOrg1,
      },
    },
  },
  {
    request: {
      query: GetPatientsByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility1.id,
        orgExternalId: mockOrg1.externalId,
        pageNumber: 0,
        pageSize: 20,
      },
    },
    result: {
      data: {
        patients: [mockPatient1, mockPatient2],
      },
    },
  },
  {
    request: {
      query: GetPatientsByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility1.id,
        orgExternalId: mockOrg1.externalId,
        pageNumber: 0,
        pageSize: 20,
      },
    },
    result: {
      data: {
        patients: [mockPatient1, mockPatient2],
      },
    },
  },
  {
    request: {
      query: GetPatientsByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility2.id,
        orgExternalId: mockOrg1.externalId,
        pageNumber: 0,
        pageSize: 20,
      },
    },
    result: {
      data: {
        patients: [],
      },
    },
  },
  {
    request: {
      query: GetPatientsByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility2.id,
        orgExternalId: mockOrg1.externalId,
        pageNumber: 0,
        pageSize: 20,
      },
    },
    result: {
      data: {
        patients: [],
      },
    },
  },
  {
    request: {
      query: GetPatientsCountByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility1.id,
        orgExternalId: mockOrg1.externalId,
      },
    },
    result: {
      data: {
        patientsCount: 2,
      },
    },
  },
  {
    request: {
      query: GetPatientsCountByFacilityWithOrgDocument,
      variables: {
        archivedStatus: ArchivedStatus.Archived,
        facilityId: mockFacility2.id,
        orgExternalId: mockOrg1.externalId,
      },
    },
    result: {
      data: {
        patientsCount: 0,
      },
    },
  },
];

describe("unarchive patient", () => {
  const mockStore = createMockStore([]);
  const mockedStore = mockStore({ facilities: [] });
  let setSelectedOrgId: jest.Mock;
  let setSelectedFacilityId: jest.Mock;
  let setFacilities: jest.Mock;
  let setArchivedPatientsCount: jest.Mock;
  let setArchivedPatients: jest.Mock;

  beforeEach(() => {
    setSelectedOrgId = jest.fn();
    setSelectedFacilityId = jest.fn();
    setFacilities = jest.fn();
    setArchivedPatientsCount = jest.fn();
    setArchivedPatients = jest.fn();
  });
  it("validates search form", async () => {
    let mocks = [
      {
        request: {
          query: GetOrganizationWithFacilitiesDocument,
          variables: {
            id: mockOrg1.internalId,
          },
        },
        result: {
          data: {
            organization: mockOrg1,
          },
        },
      },
    ];
    render(
      <Provider store={mockedStore}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <UnarchivePatient
            organizations={mockOrgs}
            currentPage={1}
            selectedFacilityId={""}
            setSelectedFacilityId={setSelectedFacilityId}
            selectedOrgId={""}
            setSelectedOrgId={setSelectedOrgId}
            archivedPatientsCount={undefined}
            setArchivedPatientsCount={setArchivedPatientsCount}
            facilities={[mockFacility1, mockFacility2]}
            setFacilities={setFacilities}
            archivedPatients={undefined}
            setArchivedPatients={setArchivedPatients}
          />
        </MockedProvider>
      </Provider>
    );
    await clickSearch();
    expect(screen.getByText("Organization is required")).toBeInTheDocument();
    expect(
      screen.getByText("Testing facility is required")
    ).toBeInTheDocument();
    await selectDropdown("Organization *", mockOrg1.name);
    expect(setArchivedPatients).toHaveBeenCalledTimes(1);
    expect(setArchivedPatients).toHaveBeenCalledWith(undefined);
    expect(setArchivedPatientsCount).toHaveBeenCalledTimes(1);
    expect(setArchivedPatientsCount).toHaveBeenCalledWith(undefined);
    expect(
      screen.queryByText("Organization is required")
    ).not.toBeInTheDocument();
    expect(
      screen.getByText("Testing facility is required")
    ).toBeInTheDocument();
    await selectDropdown("Testing facility *", mockFacility1.name);
    expect(
      screen.queryByText("Testing facility is required")
    ).not.toBeInTheDocument();
    // selecting a default option clears
    await selectDropdown("Testing facility *", "Select a testing facility");
    expect(setArchivedPatients).toHaveBeenNthCalledWith(2, undefined);
    expect(setArchivedPatientsCount).toHaveBeenNthCalledWith(2, undefined);
    expect(
      screen.getByText(
        "Filter by organization and testing facility to display archived patients."
      )
    ).toBeInTheDocument();
  });
  it("displays archived patients with pagination", async () => {
    const { container } = render(
      <Provider store={mockedStore}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <MemoryRouter>
            <UnarchivePatient
              organizations={mockOrgs}
              currentPage={1}
              selectedFacilityId={mockFacility1.id}
              setSelectedFacilityId={setSelectedFacilityId}
              selectedOrgId={mockOrg1.internalId}
              setSelectedOrgId={setSelectedOrgId}
              archivedPatientsCount={2}
              setArchivedPatientsCount={setArchivedPatientsCount}
              facilities={[mockFacility1, mockFacility2]}
              setFacilities={setFacilities}
              archivedPatients={[mockPatient1, mockPatient2]}
              setArchivedPatients={setArchivedPatients}
            />
          </MemoryRouter>
        </MockedProvider>
      </Provider>
    );
    await selectDropdown("Organization *", mockOrg1.name);
    expect(setArchivedPatients).toHaveBeenCalledWith(undefined);
    expect(setArchivedPatientsCount).toHaveBeenCalledWith(undefined);
    await selectDropdown("Testing facility *", mockFacility1.name);
    await clickSearch();
    await waitForElementToBeRemoved(() => screen.queryByText("Loading..."));
    expect(setArchivedPatients).toHaveBeenLastCalledWith(
      expect.arrayContaining([mockPatient1, mockPatient2])
    );
    expect(setArchivedPatientsCount).toHaveBeenLastCalledWith(2);
    checkTableHeadersExist();
    expect(
      screen.getByText("Showing 1-2 of 2", { exact: false })
    ).toBeInTheDocument();
    let resultsRow = screen.getAllByTestId("sr-archived-patient-row");
    resultsRow.forEach((row, index) => {
      if (index === 0) {
        expect(within(row).getByText("Gutmann, Rod")).toBeInTheDocument();
        expect(within(row).getByText("05/19/1927")).toBeInTheDocument();
        expect(within(row).getByText("All facilities")).toBeInTheDocument();
      } else if (index === 1) {
        expect(within(row).getByText("Mode, Mia")).toBeInTheDocument();
        expect(within(row).getByText("11/20/1973")).toBeInTheDocument();
        expect(within(row).getByText("Mars Facility")).toBeInTheDocument();
      }
    });
    expect(container).toMatchSnapshot();
  });
  it("displays no results", async () => {
    render(
      <Provider store={mockedStore}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <MemoryRouter>
            <UnarchivePatient
              organizations={mockOrgs}
              currentPage={1}
              selectedFacilityId={mockFacility2.id}
              setSelectedFacilityId={setSelectedFacilityId}
              selectedOrgId={mockOrg1.internalId}
              setSelectedOrgId={setSelectedOrgId}
              archivedPatientsCount={0}
              setArchivedPatientsCount={setArchivedPatientsCount}
              facilities={[mockFacility1, mockFacility2]}
              setFacilities={setFacilities}
              archivedPatients={[]}
              setArchivedPatients={setArchivedPatients}
            />
          </MemoryRouter>
        </MockedProvider>
      </Provider>
    );
    await selectDropdown("Organization *", mockOrg1.name);
    await selectDropdown("Testing facility *", mockFacility2.name);
    await act(async () => await userEvent.click(screen.getByText("Search")));
    await waitForElementToBeRemoved(() => screen.queryByText("Loading..."));
    checkTableHeadersExist();
    expect(screen.getByText("No results")).toBeInTheDocument();
    expect(
      screen.queryByText("Showing", { exact: false })
    ).not.toBeInTheDocument();
    expect(setArchivedPatients).toHaveBeenLastCalledWith(
      expect.arrayContaining([])
    );
    expect(setArchivedPatientsCount).toHaveBeenLastCalledWith(0);
  });
});
const clickSearch = async () => {
  await act(async () => await userEvent.click(screen.getByText("Search")));
};
const checkTableHeadersExist = () => {
  expect(screen.getByText("Patient")).toBeInTheDocument();
  expect(screen.getByText("Date of birth")).toBeInTheDocument();
  expect(screen.getByText("Action")).toBeInTheDocument();
};
