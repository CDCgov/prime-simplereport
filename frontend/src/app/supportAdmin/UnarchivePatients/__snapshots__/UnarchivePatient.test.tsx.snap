// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unarchive patient displays archived patients with pagination 1`] = `
<div>
  <div
    class="prime-home flex-1"
  >
    <div
      class="grid-container"
    >
      <div
        class="grid-row"
      >
        <div
          class="prime-container card-container"
        >
          <div
            class="usa-card__header"
          >
            <div
              class="display-flex flex-align-center"
            >
              <h1
                class="font-sans-lg margin-y-0"
              >
                Unarchive patient
              </h1>
              <span
                class="sr-showing-patients-on-page margin-left-4"
              >
                Loading...
                Showing 
                1
                -
                2
                 
                of 
                2
              </span>
            </div>
          </div>
          <div
            class="bg-base-lightest"
          >
            <form
              class="display-flex grid-row grid-gap flex-row flex-align-end padding-x-3 padding-y-2"
              role="search"
            >
              <div
                class="desktop:grid-col-4 mobile:grid-col-12"
              >
                <div
                  class="usa-form-group prime-dropdown "
                >
                  <label
                    class="usa-label"
                    for="7"
                  >
                    Organization
                    <abbr
                      class="usa-hint usa-hint--required"
                      title="required"
                    >
                       
                      *
                    </abbr>
                  </label>
                  <select
                    aria-required="true"
                    class="usa-select"
                    id="7"
                    name="organization"
                  >
                    <option
                      value=""
                    >
                      Select an organization
                    </option>
                    <option
                      aria-selected="true"
                      value="f34183c4-b4c5-449f-98b0-2e02abb7aae0"
                    >
                      Space Org
                    </option>
                  </select>
                </div>
              </div>
              <div
                class="desktop:grid-col-4 mobile:grid-col-12"
              >
                <div
                  class="usa-form-group prime-dropdown "
                >
                  <label
                    class="usa-label"
                    for="8"
                  >
                    Testing facility
                    <abbr
                      class="usa-hint usa-hint--required"
                      title="required"
                    >
                       
                      *
                    </abbr>
                  </label>
                  <select
                    aria-required="true"
                    class="usa-select"
                    id="8"
                    name="facility"
                  >
                    <option
                      value=""
                    >
                      Select a testing facility
                    </option>
                    <option
                      aria-selected="true"
                      value="bc0536e-4564-4291-bbf3-0e7b0731f9e8"
                    >
                      Mars Facility
                    </option>
                    <option
                      aria-selected="false"
                      value="d70bb3b3-96bd-40d1-a3ce-b266a7edb91d"
                    >
                      Jupiter Facility
                    </option>
                  </select>
                </div>
              </div>
              <div
                class="mobile:margin-top-3"
              >
                <button
                  class="usa-button"
                  type="submit"
                >
                  Search
                </button>
              </div>
            </form>
          </div>
          <div
            class="usa-card__body sr-patient-list"
          >
            <table
              class="usa-table usa-table--borderless width-full"
            >
              <thead>
                <tr>
                  <th
                    scope="col"
                  >
                    Patient
                  </th>
                  <th
                    scope="col"
                  >
                    Date of birth
                  </th>
                  <th
                    scope="col"
                  >
                    Testing facility
                  </th>
                  <th
                    scope="col"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                aria-live="polite"
              >
                <tr
                  data-testid="sr-archived-patient-row"
                >
                  <td>
                    Gutmann, Rod
                  </td>
                  <td>
                    05/19/1927
                  </td>
                  <td>
                    All facilities
                  </td>
                  <td>
                    <button
                      class="usa-button"
                      type="button"
                    >
                      Unarchive
                    </button>
                  </td>
                </tr>
                <tr
                  data-testid="sr-archived-patient-row"
                >
                  <td>
                    Mode, Mia
                  </td>
                  <td>
                    11/20/1973
                  </td>
                  <td>
                    Mars Facility
                  </td>
                  <td>
                    <button
                      class="usa-button"
                      type="button"
                    >
                      Unarchive
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            class="usa-card__footer"
          >
            <nav
              aria-label="Pagination"
              class="usa-pagination"
              role="navigation"
            >
              <ol>
                <li>
                  <a
                    aria-label="Page 1"
                    class="is-active"
                    href="/admin/unarchive-patient/1undefined"
                  >
                    <span>
                      1
                    </span>
                  </a>
                </li>
              </ol>
            </nav>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
